pkg/operation/botanist/etcdencryption.go

```diff
@@ -36,85 +38,197 @@ import (
+
+			secretsEncryptionStatus := &encryptionconfiguration.ETCDEncryptionState{
+				ForcePlainTextResources: forcePlaintextSecrets,
+				IsEncryptionActive:      isEncryptionActive,
+			}
+
+			infodata.UpsertInfoData(&gardenerResourceDataList, common.ETCDSecretsEncryptionStateName, secretsEncryptionStatus)
+		} else {
+			b.Logger.Infof("Encryption state for etcd secrets is already present in ShootState %s/%s", shootState.Namespace, shootState.Name)
+		}
+
+		shootState.Spec.Gardener = gardenerResourceDataList
+		return nil
+	})
 	if err != nil {
 		return err
 	}
-
-	return b.syncEncryptionConfigurationToGarden(ctx, secret)
+	b.ShootState = shootState
+	return nil
 }
 
-func (b *Botanist) syncEncryptionConfigurationFromGardenCluster(ctx context.Context) error {
-	gardenSecret := &corev1.Secret{}
-	if err := b.K8sGardenClient.Client().Get(ctx, common.GardenEtcdEncryptionSecretKey(b.Shoot.Info.Namespace, b.Shoot.Info.Name), gardenSecret); err != nil {
+// SyncForcePlaintextSecrets syncs the annotation which determines whether secret resources should be kept unencrypted in the
+// Shoot's ETCD from the ETCD encryption configuration secret.
+func (b *Botanist) SyncForcePlaintextSecrets(ctx context.Context) error {
+	secret := &corev1.Secret{ObjectMeta: kutil.ObjectMeta(b.Shoot.SeedNamespace, common.EtcdEncryptionSecretName)}
+	err := b.K8sSeedClient.Client().Get(ctx, kutil.KeyFromObject(secret), secret)
+	if err != nil {
 		if apierrors.IsNotFound(err) {
 			return nil
 		}
 		return err
 	}
+	shootState := &gardencorev1alpha1.ShootState{ObjectMeta: kutil.ObjectMeta(b.Shoot.Info.Namespace, b.Shoot.Info.Name)}
+	_, err = controllerutil.CreateOrUpdate(ctx, b.K8sGardenClient.Client(), shootState, func() error {
+		gardenerResourceDataList := gardencorev1alpha1helper.GardenerResourceDataList(shootState.Spec.Gardener)
+		forcePlaintextSecrets := kutil.HasMetaDataAnnotation(secret, common.EtcdEncryptionForcePlaintextAnnotationName, "true")
+		encryptionStateInfoData, err := infodata.GetInfoData(gardenerResourceDataList, common.ETCDSecretsEncryptionStateName)
+		if err != nil {
+			return err
+		}
+		etcdEncryptionState, ok := encryptionStateInfoData.(*encryptionconfiguration.ETCDEncryptionState)
+		if !ok {
+			return fmt.Errorf("Could not convert GardenerResourceData entry %s to encryptionconfiguration.ETCDEncryptionState", common.ETCDSecretsEncryptionStateName)
+		}
+		etcdEncryptionState.ForcePlainTextResources = forcePlaintextSecrets
+		err = infodata.UpsertInfoData(&gardenerResourceDataList, common.ETCDSecretsEncryptionStateName, etcdEncryptionState)
+		if err != nil {
+			return err
+		}
+		shootState.Spec.Gardener = gardenerResourceDataList
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	b.ShootState = shootState
+	return nil
+}
+
+// GenerateEncryptionConfiguration generates data for encryption configuration and saves it in the shootstate
+func (b *Botanist) GenerateEncryptionConfiguration(ctx context.Context) error {
```

Comment:
I would not bundle the generation and the store-in-state part into a single function. Instead, one function should generate the stuff, another one takes it and puts it into the state. Then you can also get rid of all the duplication in this function (most stuff is already in the first function).